### ISSUE TITLE
Current master build fails (tests fail?)

### ISSUE THREAD

**Issue by mouse07410** [agda/agda#4765]
MacOS Catalina 10.15.5, Xcode-11.5, GHC-8.8.3, Cabal-3.2.0.0 (Stack-2.3.1), current Agda master.

Looks like the main binary `agda` does build,  but `agda-tests` doesn't.

```
$ time make install test 2>&1 | tee make-out.txt
. . . . .
[62 of 62] Compiling Main             ( test/Main.hs, dist-2.6.2/build/agda-tests/agda-tests-tmp/Main.dyn_o )
Linking ./dist-2.6.2/build/agda-tests/agda-tests ...
Undefined symbols for architecture x86_64:
  "_Agdazm2zi6zi2zm8oFcbAqXN6KBcY6RrJJp6p_AgdaziUtilsziList_mergeStrictlyOrderedBy_info", referenced from:
      _Lc3ECI_info in Range.dyn_o
      _Lc3EBl_info in Range.dyn_o
  "_Agdazm2zi6zi2zm8oFcbAqXN6KBcY6RrJJp6p_AgdaziInteractionziHighlightingziRange_isLeftOf_closure", referenced from:
      _Lc3ECI_info in Range.dyn_o
      _Lc3EBl_info in Range.dyn_o
  "_Agdazm2zi6zi2zm8oFcbAqXN6KBcY6RrJJp6p_AgdaziSyntaxziPosition_zddmgetRangezufuse_closure", referenced from:
      _Lu32ax_srt in Position.dyn_o
  "_Agdazm2zi6zi2zm8oFcbAqXN6KBcY6RrJJp6p_AgdaziSyntaxziPosition_zddmgetRangezufuse_info", referenced from:
      _Ls2Ybk_info in Position.dyn_o
  "_Agdazm2zi6zi2zm8oFcbAqXN6KBcY6RrJJp6p_AgdaziSyntaxziConcreteziName_simpleName_info", referenced from:
      _Ls1PXJ_info in Generators.dyn_o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
`clang' failed in phase `Linker'. (Exit code: 1)
cabal: Leaving directory '.'
cabal: Error: some packages failed to install:
Agda-2.6.2-8oFcbAqXN6KBcY6RrJJp6p failed during the building phase. The
exception was:
ExitFailure 1


real	26m42.542s
user	25m1.755s
sys	0m31.980s
make: *** [install-bin] Error 1
```

Complete build log: [make-out.txt](https://github.com/agda/agda/files/4804590/make-out.txt)

**Comment by andreasabel** [comment]
Are these the same cabal settings as in #4569?

**Comment by asr** [comment]
Does `make clean` fix the problem?

**Comment by mouse07410** [comment]
> Are these the same cabal settings as in #4569?

Naturally. Consistency rules! ;-) :-)

> Does `make clean` fix the problem?

Absolutely not.

### Update
Today's merges did not affect this problem.

**Comment by mouse07410** [comment]
Also (unrelated, but...):
```
$ make user-manual-html
======================================================================
========================= User manual (HTML) =========================
======================================================================
Running Sphinx v3.1.0

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sphinx/config.py", line 319, in eval_config_file
    execfile_(filename, namespace)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sphinx/util/pycompat.py", line 88, in execfile_
    exec(code, _globals)
  File "/Users/ur20980/Src/agda/doc/user-manual/conf.py", line 80, in <module>
    import sphinx_rtd_theme
ModuleNotFoundError: No module named 'sphinx_rtd_theme'

make[1]: *** [html] Error 2
make: *** [user-manual-html] Error 2
```

**Comment by asr** [comment]
I couldn't reproduce the problem on Ubuntu (GHC 8.8.3, cabal-install 3.2.0.0) after running
```
$ make install-bin test CABAL_OPTS="--enable-shared --enable-executable-dynamic --disable-static"
```

**Comment by mouse07410** [comment]
> I couldn't reproduce the problem on Ubuntu...

One reason could be - build flags do not seem to affect all the dependent packages.

To ensure your environment is "reasonably" similar, you'd ned to *build the whole Agda*  (probably a-fresh, or over a clean installation - with *no* dependent libraries lingering in the `~/.cabal/store/...`) using `~/.cabal/config` with my settings:
```yaml
-- library-profiling:
shared: True
-- static:
executable-dynamic: True
-- executable-static: False
-- profiling:
-- executable-profiling:
-- profiling-detail:
-- library-profiling-detail:
optimization: True
-- debug-info: False
-- library-for-ghci:
-- split-sections: False
-- split-objs: False
-- executable-stripping:
library-stripping: False
-- configure-option:
-- user-install: True
. . . . .
-- tests: False
-- coverage: False
-- library-coverage:
-- exact-configuration: False
-- benchmarks: False
-- relocatable: False
-- response-files:
-- allow-depending-on-private-libs:
-- cabal-lib-version:
-- constraint:
-- preference:
-- solver: modular
-- allow-older: False
-- allow-newer: False
-- write-ghc-environment-files:
documentation: True
-- doc-index-file: $datadir/doc/$arch-$os-$compiler/index.html
```

~~I thought we've been through this in #4569?~~
1. We had somewhat related discussion: https://github.com/agda/agda/issues/4569#issuecomment-610201757
2. It looks like tests (who failed to build in https://github.com/agda/agda/issues/4569#issuecomment-619977007) were never fixed? Or am I missing/skipping something?

**Comment by asr** [comment]
> To ensure your environment is "reasonably" similar, you'd ned to _build the whole Agda_ (probably a-fresh, or over a clean installation - with _no_ dependent libraries lingering in the `~/.cabal/store/...`) using `~/.cabal/config` with my settings:

If we are using Cabal v1-commands why are you mentioning `~/.cabal/store/...`?

If you attach your `~/.cabal/config` file I'll use it via the `--config-file` Cabal option.

**Comment by mouse07410** [comment]
> If we are using Cabal v1-commands why are you mentioning `~/.cabal/store/...`?

You're right, my mistake. But all the more reasons to experiment with a clean/fresh installation, as v1- commands are certain to leave stuff where the subsequent invocations would find it.

> If you attach your `~/.cabal/config` file I'll use it via the `--config-file` Cabal option.

Here it is - you might want to go over it for things that must be set, but refer to my home directory. [cabal-config.txt](https://github.com/agda/agda/files/4814396/cabal-config.txt)

And I want to emphasize again the need for a *fresh* installation, so you can be sure the build does not just re-use what was already compiled/installed by the previous successful runs with a different configuration.

**Comment by asr** [comment]
I'm testing in a fresh environment (fresh GHC, fresh Cabal and fresh Agda). Can you reproduce the issue setting `jobs: 1` in the `config` file?

**Comment by asr** [comment]
I couldn't reproduce the problem using your `config` file  (tested with `jobs: 1` and `jobs: 4`).

**Comment by mouse07410** [comment]
Hmm, funny... I changed to `jobs: 1`, and the entire build completed fine. Only several tests failed.
```
$ make install test 2>&1 | tee make-out6.txt
touch src/full/Agda/VersionCommit.hs
===================== Installing using Cabal with test suites ============
time cabal v1-install --disable-documentation --builddir=./dist-2.6.2 --enable-tests --disable-library-profiling -fenable-cluster-counting --ghc-options="+RTS -M3G -RTS" 
Resolving dependencies...
In order, the following will be installed:
Agda-2.6.2 *test (reinstall) (changes: QuickCheck-2.13.2 added,
filemanip-0.3.6.3 added, process-extras-0.7.4 added, tasty-1.2.3 added,
tasty-hunit-0.10.0.2 added, tasty-quickcheck-0.10.1.1 added,
tasty-silver-3.1.13 added, temporary-1.3 added, unix-compat-0.5.2 added)
Warning: Note that reinstalls are always dangerous. Continuing anyway...
[1 of 1] Compiling Main             ( dist-2.6.2/setup/setup.hs, dist-2.6.2/setup/Main.o )
Linking ././dist-2.6.2/setup/setup ...
Configuring Agda-2.6.2...
Preprocessing executable 'agda-mode' for Agda-2.6.2..
Building executable 'agda-mode' for Agda-2.6.2..
[1 of 2] Compiling Paths_Agda       ( dist-2.6.2/build/agda-mode/autogen/Paths_Agda.hs, dist-2.6.2/build/agda-mode/agda-mode-tmp/Paths_Agda.dyn_o )
[2 of 2] Compiling Main             ( src/agda-mode/Main.hs, dist-2.6.2/build/agda-mode/agda-mode-tmp/Main.dyn_o )
Linking ./dist-2.6.2/build/agda-mode/agda-mode ...
Preprocessing library for Agda-2.6.2..
unused rules: 6
Building library for Agda-2.6.2..
[  1 of 371] Compiling Agda.Interaction.ExitCode ( src/full/Agda/Interaction/ExitCode.hs, dist-2.6.2/build/Agda/Interaction/ExitCode.o )
[  2 of 371] Compiling Agda.Interaction.Options.IORefs ( src/full/Agda/Interaction/Options/IORefs.hs, dist-2.6.2/build/Agda/Interaction/Options/IORefs.o )
[  3 of 371] Compiling Agda.Termination.CutOff ( src/full/Agda/Termination/CutOff.hs, dist-2.6.2/build/Agda/Termination/CutOff.o )
. . . . .
[371 of 371] Compiling Agda.Main        ( src/full/Agda/Main.hs, dist-2.6.2/build/Agda/Main.o )
Preprocessing executable 'agda' for Agda-2.6.2..
Building executable 'agda' for Agda-2.6.2..
[1 of 1] Compiling Main             ( src/main/Main.hs, dist-2.6.2/build/agda/agda-tmp/Main.dyn_o )
Linking ./dist-2.6.2/build/agda/agda ...
Preprocessing test suite 'agda-tests' for Agda-2.6.2..
Building test suite 'agda-tests' for Agda-2.6.2..
[ 1 of 62] Compiling Internal.Helpers ( test/Internal/Helpers.hs, dist-2.6.2/build/agda-tests/agda-tests-tmp/Internal/Helpers.dyn_o )
[ 2 of 62] Compiling Internal.Compiler.MAlonzo.Encode ( test/Internal/Compiler/MAlonzo/Encode.hs, dist-2.6.2/build/agda-tests/agda-tests-tmp/Internal/Compiler/MAlonzo/Encode.dyn_o )
[ 3 of 62] Compiling Internal.Interaction.Library ( test/Internal/Interaction/Library.hs, dist-2.6.2/build/agda-tests/agda-tests-tmp/Internal/Interaction/Library.dyn_o )
[ 4 of 62] Compiling Internal.Syntax.Common ( test/Internal/Syntax/Common.hs, dist-2.6.2/build/agda-tests/agda-tests-tmp/Internal/Syntax/Common.dyn_o )
. . . . .
[62 of 62] Compiling Main             ( test/Main.hs, dist-2.6.2/build/agda-tests/agda-tests-tmp/Main.dyn_o )
Linking ./dist-2.6.2/build/agda-tests/agda-tests ...
Installing executable agda-mode in /Users/ur20980/.cabal/bin
Installing library in /Users/ur20980/.cabal/lib/x86_64-osx-ghc-8.8.3/Agda-2.6.2-8PNdkEVt3Wv8ObSy1XoUVq
Installing executable agda in /Users/ur20980/.cabal/bin
Generating Agda library interface files...
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Bool.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Char.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Char/Properties.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Coinduction.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Cubical/Path.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Cubical/Id.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Cubical/Sub.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Cubical/Glue.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Cubical/HCompU.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Equality.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Equality/Erase.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Equality/Rewrite.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Float.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Float/Properties.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/FromNat.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/FromNeg.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/FromString.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/IO.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Int.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/List.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Maybe.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Nat.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Reflection.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Reflection/Properties.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Sigma.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Size.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Strict.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/String.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/String/Properties.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/TrustMe.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Unit.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Word.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Builtin/Word/Properties.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Primitive.agda
... /Users/ur20980/src/agda/src/data/lib/prim/Agda/Primitive/Cubical.agda
Installing executable agda-mode in /Users/ur20980/.cabal/bin
Installing library in /Users/ur20980/.cabal/lib/x86_64-osx-ghc-8.8.3/Agda-2.6.2-8PNdkEVt3Wv8ObSy1XoUVq
Installing executable agda in /Users/ur20980/.cabal/bin
Completed    Agda-2.6.2

real    26m34.914s
user    24m50.599s
sys     0m31.143s
/Users/ur20980/src/agda/dist-2.6.2/build/agda-mode/agda-mode compile
Loading quail/latin-ltx...
. . . . .
It seems as if setup has already been performed.
git submodule update --init src/fix-whitespace
cd src/fix-whitespace && cabal v1-install
Resolving dependencies...
Configuring filepattern-0.1.2...
Preprocessing library for filepattern-0.1.2..
Building library for filepattern-0.1.2..
[1 of 8] Compiling System.FilePattern.ListBy ( src/System/FilePattern/ListBy.hs, dist/build/System/FilePattern/ListBy.o )
. . . . .
[8 of 8] Compiling System.FilePattern ( src/System/FilePattern.hs, dist/build/System/FilePattern.o )
Preprocessing library for filepattern-0.1.2..
Running Haddock on library for filepattern-0.1.2..
Warning: The documentation for the following packages are not installed. No
links will be generated to these packages: clock-0.8
Haddock coverage:
Warning: '<.>' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: '</>' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
 100% (  5 /  5) in 'System.FilePattern.Directory'
 100% ( 12 / 12) in 'System.FilePattern'
Documentation created: dist/doc/html/filepattern/index.html
Installing library in /Users/ur20980/.cabal/lib/x86_64-osx-ghc-8.8.3/filepattern-0.1.2-FjzLrZxpgX04Xq90pUmrnh
Completed    filepattern-0.1.2
Configuring fix-whitespace-0.0.5...
Preprocessing executable 'fix-whitespace' for fix-whitespace-0.0.5..
Building executable 'fix-whitespace' for fix-whitespace-0.0.5..
[1 of 2] Compiling ParseConfig      ( ParseConfig.hs, dist/build/fix-whitespace/fix-whitespace-tmp/ParseConfig.dyn_o )
[2 of 2] Compiling Main             ( FixWhitespace.hs, dist/build/fix-whitespace/fix-whitespace-tmp/Main.dyn_o )
Linking dist/build/fix-whitespace/fix-whitespace ...
Warning: No documentation was generated as this package does not contain a
library. Perhaps you want to use the --executables, --tests, --benchmarks or
--foreign-libraries flags.
Installing executable fix-whitespace in /Users/ur20980/.cabal/bin
Completed    fix-whitespace-0.0.5
Updating documentation index
/Users/ur20980/.cabal/share/doc/x86_64-osx-ghc-8.8.3/index.html
Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default
Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default
src/fix-whitespace/dist/build/fix-whitespace/fix-whitespace --check
======================================================================
=========== Suite of successful tests: mini-library Common ===========
======================================================================
Bool.agda
Char.agda
. . . . .
Writing interface file /Users/ur20980/src/agda/examples/uptodate/A.agdai.
Wrote interface file.
Accumulated statistics.
Finished A.
  New module. Let's check it out.
  current options  = [--safe:  False, --without-K:  False,
                      --no-universe-polymorphism:  False, --no-sized-types:  False,
                      --no-guardedness:  False, --no-subtyping:  True,
                      --no-cumulativity:  True, --cubical:  False, --prop:  False]
  imported options = [--safe:  False, --without-K:  False,
                      --no-universe-polymorphism:  False, --no-sized-types:  False,
                      --no-guardedness:  False, --no-subtyping:  True,
                      --no-cumulativity:  True, --cubical:  False, --prop:  False]
  Now we've looked at A
Testing vfl/Typechecker.agda... 
ok
======================================================================
======================= Standard library test ========================
======================================================================
Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default

GenerateEverything.hs:12:1: error:
    Could not load module ‘System.FilePath.Find’
    It is a member of the hidden package ‘filemanip-0.3.6.3’.
    You can run ‘:set -package filemanip’ to expose it.
    (Note: this unloads all the modules in the current scope.)
    It is a member of the hidden package ‘filemanip-0.3.6.3’.
    You can run ‘:set -package filemanip’ to expose it.
    (Note: this unloads all the modules in the current scope.)
    It is a member of the hidden package ‘filemanip-0.3.6.3’.
    You can run ‘:set -package filemanip’ to expose it.
    (Note: this unloads all the modules in the current scope.)
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
   |
12 | import System.FilePath.Find
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
make: *** [std-lib-test] Error 1
```

Complete log: [make-out6.txt](https://github.com/agda/agda/files/4815543/make-out6.txt)

So, what gives? I was planning to change `jobs: 12` to `jobs: 20` (upgraded CPU) - looks like for Agda I have to keep it at 4 or below? Perhaps, make it a project-local restriction?

Also, the failed tests? Should I worry?

**Comment by asr** [comment]
The tests shouldn't fail. Please run 

```
$ make std-lib
$ make cubical
```

and run again the tests.

**Comment by asr** [comment]
You could avoid the two above instructions cloning the repository using https://github.com/agda/agda/blob/f95313eb004c29b6d9cf85d60bd3b9691486bd55/HACKING.md#L46

**Comment by mouse07410** [comment]
> You could avoid the two above instructions cloning the repository using
> `git clone --recurse-submodules git@github.com:agda/agda.git`

Alas, that's how I cloned. And I pull updates with `git pull --recurse-submodules && git submodule update --init --recursive`.

A lot of tests pass - but quite a few seem to fail. I'll post the results tomorrow.

### Update
Tests keep failing. With a lot of messages like
```
> Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default
bash: Issue2728-3.agda: command not found
```

Here's the log: [make-test1.txt](https://github.com/agda/agda/files/4816527/make-test1.txt)

Also, I looked at the file "it" (Agda? `agda-tests`?) seems to load `$HOME/.ghc/x86_64-darwin-8.8.3/environments/default` - and was surprised to find this at the end: `package-id Agd-2.6.1-003b72f0`. I suspect it's a relict from one of the previous installations (or attempts to install). Here's the complete file: [default.txt](https://github.com/agda/agda/files/4816520/default.txt)  
I suspect I need to get rid of that Agda-2.6.1? Could you help recommending how to do it? In case it matters:
```
$ agda --version
Agda version 2.6.2-91763e8-dirty
$ type -all agda
agda is /Users/ur20980/.cabal/bin/agda
agda is /Users/ur20980/.local/bin/agda
$ ll ~/.local/bin/agda
lrwxr-xr-x  1 ur20980  MITLL\Domain Users  21 Jun 22 16:18 /Users/ur20980/.local/bin/agda@ -> ../../.cabal/bin/agda
```

**Comment by asr** [comment]
Which is the output of
```
ghc-pkg list | grep Agda*
```
?

**Comment by mouse07410** [comment]
```sh
$ ghc-pkg list | grep Agda
    Agda-2.6.2
$ ghc-pkg list | grep Agda*
$ ghc-pkg list | grep "Agda*"
    Agda-2.6.2
$
```

**Comment by asr** [comment]
It seems the problem is the output

```
Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default
```

1. I don't know when GHC generates this output.
2. In `make interaction` we run GHC with the flag `--ghc-arg=-v0`. From https://gitlab.haskell.org/ghc/ghc/-/issues/16879 GHC shouldn't show the above output using that flag, but I don't know if GHC 8.8.3 includes the fix to that issue.

Which is the output of
```bash
$ cd test/interaction
$ make Issue2728.cmp  
```
?

**Comment by mouse07410** [comment]
```
$ cd test/interaction/
$ make Issue2728.cmp
Issue2728
0a1
> Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default
bash: Issue2728.agda: command not found
=== Old output ===
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue2728 (Issue2728.agda). " t)
(agda2-info-action "*Type-checking*" "Issue2728.agda:2,11-14 The following names are not declared in the same scope as their syntax or fixity declaration (i.e., either not in scope at all, imported from another module, or declared in a super module): _+_ when scope checking the declaration record R where infix 0 _+_ " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Warnings*" "Issue2728.agda:2,11-14 The following names are not declared in the same scope as their syntax or fixity declaration (i.e., either not in scope at all, imported from another module, or declared in a super module): _+_ when scope checking the declaration record R where infix 0 _+_ " nil)
((last . 1) . (agda2-goals-action '()))
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue2728 (Issue2728.agda). " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
=== New output ===
Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue2728 (Issue2728.agda). " t)
(agda2-info-action "*Type-checking*" "Issue2728.agda:2,11-14 The following names are not declared in the same scope as their syntax or fixity declaration (i.e., either not in scope at all, imported from another module, or declared in a super module): _+_ when scope checking the declaration record R where infix 0 _+_ " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Warnings*" "Issue2728.agda:2,11-14 The following names are not declared in the same scope as their syntax or fixity declaration (i.e., either not in scope at all, imported from another module, or declared in a super module): _+_ when scope checking the declaration record R where infix 0 _+_ " nil)
((last . 1) . (agda2-goals-action '()))
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue2728 (Issue2728.agda). " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
=== Diff ===
{+Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default+}
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue2728 (Issue2728.agda). " t)
(agda2-info-action "*Type-checking*" "Issue2728.agda:2,11-14 The following names are not declared in the same scope as their syntax or fixity declaration (i.e., either not in scope at all, imported from another module, or declared in a super module): _+_ when scope checking the declaration record R where infix 0 _+_ " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Warnings*" "Issue2728.agda:2,11-14 The following names are not declared in the same scope as their syntax or fixity declaration (i.e., either not in scope at all, imported from another module, or declared in a super module): _+_ when scope checking the declaration record R where infix 0 _+_ " nil)
((last . 1) . (agda2-goals-action '()))
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue2728 (Issue2728.agda). " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
Accept new error [y/n/Q] (default: no)? 
```
I did not continue from here (typed Q).

**Comment by asr** [comment]
Please remove the directory `/Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/` and try again.

**Comment by mouse07410** [comment]
> Please remove the directory `/Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/` and try again

Considering that Agda is not the main part of my GHC and Haskell adventure, I'm a little hesitant to employ such a drastic measure. 

Do you know who/what app created and filled that directory? Here's it's current contents:
```
$ ll /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments
total 8
drwxr-xr-x  3 ur20980  MITLL\Domain Users   96 Jun 22 22:41 ./
drwxr-xr-x  4 ur20980  MITLL\Domain Users  128 Apr  5 21:49 ../
-rw-r--r--  1 ur20980  MITLL\Domain Users  730 Jun 22 22:41 default
$ cat /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default 
clear-package-db
global-package-db
package-db /Users/ur20980/.cabal/store/ghc-8.8.3/package.db
package-id ghc-8.8.3
package-id bytestring-0.10.10.0
package-id unix-2.7.2.2
package-id base-4.13.0.0
package-id time-1.9.3
package-id hpc-0.6.0.3
package-id filepath-1.4.2.1
package-id process-1.6.8.0
package-id array-0.5.4.0
package-id integer-gmp-1.0.2.0
package-id containers-0.6.2.1
package-id ghc-boot-8.8.3
package-id binary-0.8.7.0
package-id ghc-prim-0.5.3
package-id ghci-8.8.3
package-id rts
package-id terminfo-0.4.1.4
package-id transformers-0.5.6.2
package-id deepseq-1.4.4.0
package-id ghc-boot-th-8.8.3
package-id pretty-1.1.3.6
package-id template-haskell-2.15.0.0
package-id directory-1.3.6.0
package-id text-1.2.4.0
```

In fact, this is what worries me:
```
bash: Issue2728.agda: command not found
```
It tells me that there probably is some kind of variable that was supposed to represent a command. Guessing, maybe something like `AGDA=$HOME/.cabal/bin/agda`. And that variable ends up being empty. So, you get something like `${AGDA} Issue2728.agda` with `AGDA=""`. So, `bash` complains.

I can't figure in the maze of Makefile and configs what that variable could be, or where to look for it. But this guess seems reasonable enough.

**Comment by asr** [comment]
For testing you can rename it instead. I think the directory is created by Cabal v2-commands. I still have some problems when combining Cabal v1- and v2-commands.

**Comment by mouse07410** [comment]
How about this:
```bash
$ fgr AGDA Makefile
6:AGDA_OPT=--no-default-libraries
65:         | $(AGDA_BIN) -v0 -i . -i .. --interaction --ignore-interfaces $(AGDA_OPT) $(RTS_$*) \
68:    then /usr/bin/env runhaskell --ghc-arg=-v0 --ghc-arg=-w ./$*.hs $(AGDA_BIN) 2>&1 | $(filter) ; \
69:    else /usr/bin/env bash ./$*.sh $(AGDA_BIN) > $(TMPDIR)/$*.tmp_out ; \
78:setup_Auto = $(AGDA_BIN) -v0 Auto/Prelude.agda
$ export AGDA_BIN=/Users/ur20980/.cabal/bin/agda
$ make Issue2728.cmp
Issue2728
$ 
```

**Comment by asr** [comment]
I don't know what is `fgr` and I have not this command on Ubuntu.

**Comment by mouse07410** [comment]
> I think the directory is created by Cabal v2-commands.

Possible, but I rather doubt it. Cabal tends to put its stuff in `~/.cabal/`. `ghcup` uses `~/.ghcup`, **and** `~/.ghc` - but mostly (only?) for the distro stuff, like GHC.

> I still have some problems when combining Cabal v1- and v2-commands.

I'm totally with you. 

But it would be good to move Agda to v2- build. It would make it easier to deal with dynamic builds too, I think.

#### Update
> I don't know what is `fgr` and I have not this command on Ubuntu.

My apologies - it's my alias, created for convenience:
```bash
alias fgr='fgrep -n'
```

**Comment by asr** [comment]
> But it would be good to move Agda to v2- build. It would make it easier to deal with dynamic builds too, I think.

See https://github.com/agda/agda/issues/4216. Patches are welcome! Keep in mind we need to support previous versions of Cabal.

**Comment by asr** [comment]
> Possible, but I rather doubt it. Cabal tends to put its stuff in `~/.cabal/`. `ghcup` uses `~/.ghcup`, **and** `~/.ghc` - but mostly (only?) for the distro stuff, like GHC.

No. When I installed a library using `cabal v2-install --lib`, Cabal adds the library to `environments/default`.

**Comment by asr** [comment]
Did you rename the `environments/` directory and run `make Issue2728.cmp`?

**Comment by mouse07410** [comment]
> Patches are welcome! 

I'm a beginner in Haskell, and even less so in Agda. Patches from me are rather unlikely at this point.

> When I installed a library using `cabal v2-install --lib`, Cabal adds the library to `.../environments/default`

Well, OK. You saw the content of my `.../default` - does it looks like anything that would/could interfere here...?

```
. . . . .
Issue1232
2,20c2,11
< Checking Issue1232.All (Issue1232/All.agda).
< Checking Issue1232.Fin (Issue1232/Fin.agda).
< Checking Issue1232.List (Issue1232/List.agda).
< Checking Issue1232.Nat (Issue1232/Nat.agda).
< (agda2-status-action "")
< (agda2-info-action "*Type-checking*" "" nil)
< (agda2-highlight-clear)
< (agda2-info-action "*Type-checking*" "Checking Issue1232 (Issue1232.agda). " t)
< (agda2-status-action "Checked")
< (agda2-info-action "*All Done*" "" nil)
< ((last . 1) . (agda2-goals-action '()))
< (agda2-status-action "")
< (agda2-info-action "*Type-checking*" "" nil)
< (agda2-highlight-clear)
< (agda2-info-action "*Type-checking*" "Checking Issue1232 (Issue1232.agda). " t)
< (agda2-status-action "Checked")
< (agda2-info-action "*All Done*" "" nil)
< ((last . 1) . (agda2-goals-action '()))
< Issue1232.hs: _build/2.6.1/agda/Issue1232.agdai: removeLink: does not exist (No such file or directory)
---
> 
> Issue1232.hs:6:1: error:
> Could not load module ‘Agda.Version’
> It is a member of the hidden package ‘Agda-2.6.1’.
> You can run ‘:set -package Agda’ to expose it.
> (Note: this unloads all the modules in the current scope.)
> Use -v (or `:set -v` in ghci) to see a list of the files searched for.
> |
> 6 | import Agda.Version
> | ^^^^^^^^^^^^^^^^^^^
bash: Issue1232.agda: command not found
=== Old output ===
. . . . .
Issue1232.hs: _build/2.6.1/agda/Issue1232.agdai: removeLink: does not exist (No such file or directory)
=== New output ===
Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default

Issue1232.hs:6:1: error:
Could not load module ‘Agda.Version’
It is a member of the hidden package ‘Agda-2.6.1’.
You can run ‘:set -package Agda’ to expose it.
(Note: this unloads all the modules in the current scope.)
Use -v (or `:set -v` in ghci) to see a list of the files searched for.
|
6 | import Agda.Version
| ^^^^^^^^^^^^^^^^^^^
=== Diff ===
Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default
[-Checking Issue1232.All (Issue1232/All.agda).
Checking Issue1232.Fin (Issue1232/Fin.agda).
Checking Issue1232.List (Issue1232/List.agda).
Checking Issue1232.Nat (Issue1232/Nat.agda).
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue1232 (Issue1232.agda). " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue1232 (Issue1232.agda). " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
Issue1232.hs: _build/2.6.1/agda/Issue1232.agdai: removeLink: does-]

{+Issue1232.hs:6:1: error:
Could+} not [-exist (No such file or directory)-] {+load module ‘Agda.Version’
It is a member of the hidden package ‘Agda-2.6.1’.
You can run ‘:set -package Agda’ to expose it.
(Note: this unloads all the modules in the current scope.)
Use -v (or `:set -v` in ghci) to see a list of the files searched for.
|
6 | import Agda.Version
| ^^^^^^^^^^^^^^^^^^^+}
Accept new error [y/n/Q] (default: no)? 
```

**Comment by asr** [comment]
I need focus in one problem at a time.

Which is the output of

```bash
$ cd test/interaction/
$ make Issue2728.cmp
```
?

**Comment by mouse07410** [comment]
> Which is the output of
> ```
> $ cd test/interaction/
> $ make Issue2728.cmp
> ```

```bash
$ type agda
agda is hashed (/Users/ur20980/.cabal/bin/agda)
$ export AGDA_BIN=/Users/ur20980/.cabal/bin/agda
$ cd ~/src/agda/test/interaction
$ make Issue2728.cmp
Issue2728
$ 
```

In other words, when I make sure `AGDA_BIN` points at something reasonable - this test seems to pass.

**Comment by asr** [comment]
The test suite uses `agda` from the build directory. From the root of the Agda repository which is the output of

```
dist-2.6.2/build/agda/agda --version
```

?

**Comment by mouse07410** [comment]
```bash
$ pwd
/Users/ur20980/src/agda
$ dist-2.6.2/build/agda/agda --version
Agda version 2.6.2-91763e8-dirty
$ ~/.cabal/bin/agda --version
Agda version 2.6.2-91763e8-dirty
$
```

Also
```bash
$ otool -L ~/.cabal/bin/agda
/Users/ur20980/.cabal/bin/agda:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
	@rpath/libHSAgda-2.6.2-8PNdkEVt3Wv8ObSy1XoUVq-ghc8.8.3.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libHSbase-4.13.0.0-ghc8.8.3.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libHSghc-prim-0.5.3-ghc8.8.3.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libHSrts_thr-ghc8.8.3.dylib (compatibility version 0.0.0, current version 0.0.0)
$ 
$ otool -L dist-2.6.2/build/agda/agda
dist-2.6.2/build/agda/agda:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
	@rpath/libHSAgda-2.6.2-8PNdkEVt3Wv8ObSy1XoUVq-ghc8.8.3.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libHSbase-4.13.0.0-ghc8.8.3.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libHSghc-prim-0.5.3-ghc8.8.3.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libHSrts_thr-ghc8.8.3.dylib (compatibility version 0.0.0, current version 0.0.0)
$
```

**Comment by asr** [comment]
And the output of
```
$ cd test/interaction/
$ make Issue2728.cmp
```
?

**Comment by asr** [comment]
Could someone using MacOS Catalina 10.15.5 confirm if `make interaction` works?

**Comment by mouse07410** [comment]
> And the output of
> ```
> $ cd test/interaction/
> $ make Issue2728.cmp
> ```

I thought I've posted it several times already?

It passes now. I think I accepted the new output with the extra string
```
Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default
```

```
$ unset AGDA_BIN
$ make Issue2728.cmp
Issue2728
$ 
```

**Comment by mouse07410** [comment]
```bash
$ pwd
/Users/ur20980/src/agda/test/interaction
$ make
Issue2728
Issue2728-2
Issue2728-3
Issue2306
Issue641
Issue3082
Issue1232
2,20c2,11
< Checking Issue1232.All (Issue1232/All.agda).
< Checking Issue1232.Fin (Issue1232/Fin.agda).
< Checking Issue1232.List (Issue1232/List.agda).
< Checking Issue1232.Nat (Issue1232/Nat.agda).
< (agda2-status-action "")
< (agda2-info-action "*Type-checking*" "" nil)
< (agda2-highlight-clear)
< (agda2-info-action "*Type-checking*" "Checking Issue1232 (Issue1232.agda). " t)
< (agda2-status-action "Checked")
< (agda2-info-action "*All Done*" "" nil)
< ((last . 1) . (agda2-goals-action '()))
< (agda2-status-action "")
< (agda2-info-action "*Type-checking*" "" nil)
< (agda2-highlight-clear)
< (agda2-info-action "*Type-checking*" "Checking Issue1232 (Issue1232.agda). " t)
< (agda2-status-action "Checked")
< (agda2-info-action "*All Done*" "" nil)
< ((last . 1) . (agda2-goals-action '()))
< Issue1232.hs: _build/2.6.1/agda/Issue1232.agdai: removeLink: does not exist (No such file or directory)
---
> 
> Issue1232.hs:6:1: error:
> Could not load module ‘Agda.Version’
> It is a member of the hidden package ‘Agda-2.6.1’.
> You can run ‘:set -package Agda’ to expose it.
> (Note: this unloads all the modules in the current scope.)
> Use -v (or `:set -v` in ghci) to see a list of the files searched for.
> |
> 6 | import Agda.Version
> | ^^^^^^^^^^^^^^^^^^^
bash: Issue1232.agda: command not found
=== Old output ===
Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default
Checking Issue1232.All (Issue1232/All.agda).
Checking Issue1232.Fin (Issue1232/Fin.agda).
Checking Issue1232.List (Issue1232/List.agda).
Checking Issue1232.Nat (Issue1232/Nat.agda).
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue1232 (Issue1232.agda). " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue1232 (Issue1232.agda). " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
Issue1232.hs: _build/2.6.1/agda/Issue1232.agdai: removeLink: does not exist (No such file or directory)
=== New output ===
Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default

Issue1232.hs:6:1: error:
Could not load module ‘Agda.Version’
It is a member of the hidden package ‘Agda-2.6.1’.
You can run ‘:set -package Agda’ to expose it.
(Note: this unloads all the modules in the current scope.)
Use -v (or `:set -v` in ghci) to see a list of the files searched for.
|
6 | import Agda.Version
| ^^^^^^^^^^^^^^^^^^^
=== Diff ===
Loaded package environment from /Users/ur20980/.ghc/x86_64-darwin-8.8.3/environments/default
[-Checking Issue1232.All (Issue1232/All.agda).
Checking Issue1232.Fin (Issue1232/Fin.agda).
Checking Issue1232.List (Issue1232/List.agda).
Checking Issue1232.Nat (Issue1232/Nat.agda).
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue1232 (Issue1232.agda). " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue1232 (Issue1232.agda). " t)
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
Issue1232.hs: _build/2.6.1/agda/Issue1232.agdai: removeLink: does-]

{+Issue1232.hs:6:1: error:
Could+} not [-exist (No such file or directory)-] {+load module ‘Agda.Version’
It is a member of the hidden package ‘Agda-2.6.1’.
You can run ‘:set -package Agda’ to expose it.
(Note: this unloads all the modules in the current scope.)
Use -v (or `:set -v` in ghci) to see a list of the files searched for.
|
6 | import Agda.Version
| ^^^^^^^^^^^^^^^^^^^+}
Accept new error [y/n/Q] (default: no)?
```

**Comment by mouse07410** [comment]
I cloned the repo a-fresh, re-doing the build. 

If there are remnants of a successful Agda-2.6.1 build - I wonder how to get rid of it, without blowing away everything I installed on this machine using Cabal.

```bash
$ ghc-pkg list | grep "Agd*"
    Agda-2.6.2
```

**Comment by asr** [comment]
> If there are remnants of a successful Agda-2.6.1 build.

Yes. The line `import Agda.Version` in `Issue1232.hs` is using the wrong version of Agda. I don't know why.

**Comment by asr** [comment]
> ```shell
> $ ghc-pkg list | grep "Agd*"
>     Agda-2.6.2
> ```

What about `ghc-pkg list --user | grep Agda`?

**Comment by mouse07410** [comment]
> What about `ghc-pkg list --user | grep Agda`?

Same. :-(

I can't find any visible traces of Agda-2.6.1. :-(

**Comment by asr** [comment]
What about `ghc-pkg find-module Agda.Version`?

**Comment by mouse07410** [comment]
```bash
$ ghc-pkg find-module Agda.Version
/Users/ur20980/.ghcup/ghc/8.8.3/lib/ghc-8.8.3/package.conf.d
    (no packages)
/Users/ur20980/.ghc/x86_64-darwin-8.8.3/package.conf.d
    Agda-2.6.2
$
```

Also:
```
$ find ~/.cabal -name '*[Aa]gd-2.6.1*' -print -exec rm -rf {} \;
/Users/ur20980/.cabal/logs/ghc-8.8.3/Agd-2.6.1-47d1e50c.log
/Users/ur20980/.cabal/logs/ghc-8.8.3/Agd-2.6.1-bd0fa011.log
/Users/ur20980/.cabal/logs/ghc-8.8.3/Agd-2.6.1-003b72f0.log
/Users/ur20980/.cabal/store/ghc-8.8.3/package.db/Agd-2.6.1-003b72f0.conf
/Users/ur20980/.cabal/store/ghc-8.8.3/lib/libHSAgd-2.6.1-003b72f0-ghc8.8.3.dylib
/Users/ur20980/.cabal/store/ghc-8.8.3/incoming/Agd-2.6.1-003b72f0.lock
$ 
```

**Comment by mouse07410** [comment]
The above did not seem to help. Somewhere there seems to remain the "ghost" of Agda-2.6.1, and somehow the tests from this build manage to find it - without letting me know *actually where* they got it from. And I'm not ready to wipe my entire Haskell installation to eradicate this module.

**Comment by asr** [comment]
What about (from https://gitlab.haskell.org/ghc/ghc/-/issues/13753)

```bash
$ export GHC_ENVIRONMENT=-
$ make Issue1232.cmp
```
?

**Comment by mouse07410** [comment]
```
$ export GHC_ENVIRONMENT=-
$ make Issue1232.cmp
Issue1232
$ 
```

### Update
```
$ export GHC_ENVIRONMENT=-
$ make Issue1232.cmp
Issue1232
$ make
Issue1232
unicodelambdasplit
noinfixdecl
empty
With-flicker
Warnings
UnderappliedConstructor
TopScope
Token-highlighting
Token-highlighting-with-parse-error
Token-highlighting-literate
Stuck
SplitResult
SplitPreserveInstanceProjection
SplitOnResultCopatternsDisabled
SplitOnHidden
. . . . .
Auto-Misc
Auto-IndexedDatatypes
Auto-EqualityReasoning
Auto-DataConstruction
Auto-BasicLogic
Allto
$
```

**Comment by mouse07410** [comment]
Tried 
```
$ cd ~/src/agda
$ make test 2>&1 | tee make-test-out.txt
. . . . .
    ltgt_QuickLaTeX:                                   OK (3.08s)
    Σ_HTML:                                            OK (0.42s)
    doc-user-manual-tools-acmart-pdflatex_LaTeX:       Golden value differs from actual value.
                                                       Accept actual value as new golden value? [yn] n
                                                       FAIL
                                                       Golden value does not match actual output.
    doc-user-manual-tools-acmart-pdflatex_QuickLaTeX:  Golden value differs from actual value.
                                                       Accept actual value as new golden value? [yn] n
                                                       FAIL
                                                       Golden value does not match actual output.
    doc-user-manual-tools-article-pdflatex_LaTeX:      Golden value differs from actual value.
                                                       Accept actual value as new golden value? [yn] n
                                                       FAIL
                                                       Golden value does not match actual output.
    doc-user-manual-tools-article-pdflatex_QuickLaTeX: Golden value differs from actual value.
                                                       Accept actual value as new golden value? [yn] n
                                                       FAIL
                                                       Golden value does not match actual output.
    doc-user-manual-tools-beamer-pdflatex_LaTeX:       Golden value differs from actual value.
                                                       Accept actual value as new golden value? [yn] n
                                                       FAIL
                                                       Golden value does not match actual output.
    doc-user-manual-tools-beamer-pdflatex_QuickLaTeX:  Golden value differs from actual value.
                                                       Accept actual value as new golden value? [yn] n
                                                       FAIL
                                                       Golden value does not match actual output.

6 out of 214 tests failed (127.56s)
make: *** [latex-html-test] Error 1
```

Looks like we've had plenty of success so far. Remaining problems seem to be with LaTeX.

Here's the complete log: [make-test-out.txt](https://github.com/agda/agda/files/4825584/make-test-out.txt)

**Comment by asr** [comment]
The "Suite of tests for the LaTeX and HTML backends" uses various LaTeX packages. We need to find the missing packages.

To know the missing LaTeX packages for the above first error, you can run:

```bash
$ cd /tmp/
$ cp your-agda-repository/doc/user-manual/tools/acmart-pdflatex.lagda.tex .
$ agda --latex --latex-dir=. acmart-pdflatex.lagda.tex
$ pdflatex acmart-pdflatex.tex
```

**Comment by mouse07410** [comment]
```
$ cd /tmp
$ cp ~/src/agda/doc/user-manual/tools/acmart-pdflatex.lagda.tex .
$ agda --latex --latex-dir=. acmart-pdflatex.lagda.tex 
Checking acmart-pdflatex (/private/tmp/acmart-pdflatex.lagda.tex).
agda.sty was not found. Copying a default version of agda.sty
into ..

$ cp ~/src/agda/src/data/agda.sty /
cp: /agda.sty: Read-only file system
$ cp ~/src/agda/src/data/agda.sty .
$ agda --latex --latex-dir=. acmart-pdflatex.lagda.tex 
$ pdflatex acmart-pdflatex.tex
This is pdfTeX, Version 3.14159265-2.6-1.40.20 (TeX Live 2019/MacPorts 2019.50896_2) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
(./acmart-pdflatex.tex
LaTeX2e <2018-12-01>

! LaTeX Error: File `acmart.cls' not found.

Type X to quit or <RETURN> to proceed,
or enter new name. (Default extension: cls)

Enter file name: 
```

**Comment by asr** [comment]
> $ cp ~/src/agda/src/data/agda.sty /
> cp: /agda.sty: Read-only file system
> $ cp ~/src/agda/src/data/agda.sty .

The above instructions are not necessaries.

> ! LaTeX Error: File `acmart.cls' not found.

You need to install the LaTeX package including the `acmart.cls` file.

**Comment by mouse07410** [comment]
![Screen Shot 2020-06-24 at 11 25 32](https://user-images.githubusercontent.com/5923577/85584804-77dca380-b60d-11ea-863c-4fdf913f0a17.png)

```bash
$ find /usr/local/texlive/2020 -name 'acmart*' -print
/usr/local/texlive/2020/texmf-dist/tex/latex/acmart
/usr/local/texlive/2020/texmf-dist/tex/latex/acmart/acmart.cls
/usr/local/texlive/2020/texmf-dist/source/latex/acmart
/usr/local/texlive/2020/texmf-dist/source/latex/acmart/acmart.ins
/usr/local/texlive/2020/texmf-dist/source/latex/acmart/acmart.dtx
/usr/local/texlive/2020/texmf-dist/bibtex/bst/acmart
/usr/local/texlive/2020/texmf-dist/doc/latex/acmart
/usr/local/texlive/2020/texmf-dist/doc/latex/acmart/acmart.pdf
/usr/local/texlive/2020/texmf-dist/doc/latex/acmart/acmart.bib
/usr/local/texlive/2020/tlpkg/tlpobj/acmart.tlpobj
/usr/local/texlive/2020/tlpkg/backups/acmart.r53899.tar.lz4
$ 
```

Somehow the installed package is not found (_by Agda_?).

Also:
```bash
$ pdflatex acmart-pdflatex.tex
This is pdfTeX, Version 3.14159265-2.6-1.40.20 (TeX Live 2019/MacPorts 2019.50896_2) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
(./acmart-pdflatex.tex
LaTeX2e <2018-12-01>

! LaTeX Error: File `acmart.cls' not found.

Type X to quit or <RETURN> to proceed,
or enter new name. (Default extension: cls)

Enter file name: /usr/local/texlive/2020/texmf-dist/tex/latex/acmart/acmart.cls
(/usr/local/texlive/2020/texmf-dist/tex/latex/acmart/acmart.cls
Document Class: acmart 2020/04/30 v1.71 Typesetting articles for the Associatio
n for Computing Machinery
(/opt/local/share/texmf-texlive/tex/latex/xkeyval/xkeyval.sty
(/opt/local/share/texmf-texlive/tex/generic/xkeyval/xkeyval.tex
(/opt/local/share/texmf-texlive/tex/generic/xkeyval/xkvutils.tex
(/opt/local/share/texmf-texlive/tex/generic/xkeyval/keyval.tex)))

! Package xkeyval Error: xkeyval loaded before \documentclass.

See the xkeyval package documentation for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.62 ...yval loaded before \protect\documentclass}
                                                  %
? 
```

**Comment by asr** [comment]
Why are you talking about of

```
This is pdfTeX, Version 3.14159265-2.6-1.40.20 (TeX Live 2019/MacPorts 2019.50896_2)
```

and 

`/usr/local/texlive/2020/` ?

That is, TeX Live 2019 and 2020? It didn't happen when I installed TeX Live on Ubuntu.

**Comment by mouse07410** [comment]
Darn... it looks like TeX Live 2019 was *also* installed by MacPorts. Addressing it as we speak.

**Comment by asr** [comment]
Note that TeX Live 2020 was released on 10 April 2020. I install a new version of TeX Live only after three months of its release. That is, I used TeX Live 2019 when running the test suite.

**Comment by mouse07410** [comment]
> Note that TeX Live 2020 was released on 10 April 2020. I install a new version of TeX Live only after three months of its release. That is, I used TeX Live 2019 when running the test suite.

Sure. But that doesn't seem to be the problem, IMHO - it's the possible conflict between what MacPorts installs (and what packages *it* makes available) vs. what I consider "my" TeX Live installation provided. I'll report as soon as the new test-run completes.

**Comment by asr** [comment]
Right, that was not the problem, but I guess no one has used TeX Live 2020 for testing.

**Comment by mouse07410** [comment]
```bash
. . . . .
    tools-compilers:                               OK (0.95s)
    tools-search-about:                            OK (0.91s)

All 46 tests passed (3.56s)
======================================================================
================= Installing the size-solver program =================
======================================================================
cabal v1-install
Resolving dependencies...
Starting     size-solver-0.2016.7.13
Building     size-solver-0.2016.7.13
Completed    size-solver-0.2016.7.13
Updating documentation index
/Users/ur20980/.cabal/share/doc/x86_64-osx-ghc-8.8.3/index.html
======================================================================
================== Testing the size-solver program ===================
======================================================================
shelltest --color --precise test/succeed.test
bash: shelltest: command not found
make[1]: *** [test] Error 127
make: *** [test-size-solver] Error 2
```

Complete log: [make-test-out.txt](https://github.com/agda/agda/files/4826877/make-test-out.txt)

**Comment by asr** [comment]
> bash: shelltest: command not found

`cabal install shelltestrunner`

**Comment by asr** [comment]
Note that you can individually run every test: https://github.com/agda/agda/blob/48423ad9bc224d30c57fe766395c5a4aadbba702/Makefile#L284-L303

**Comment by mouse07410** [comment]
```bash
$ hash -r
$ make test-size-solver
======================================================================
================= Installing the size-solver program =================
======================================================================
cabal v1-install
Resolving dependencies...
Starting     size-solver-0.2016.7.13
Building     size-solver-0.2016.7.13
Completed    size-solver-0.2016.7.13
Updating documentation index
/Users/ur20980/.cabal/share/doc/x86_64-osx-ghc-8.8.3/index.html
======================================================================
================== Testing the size-solver program ===================
======================================================================
shelltest --color --precise test/succeed.test
:test/succeed.test:1: [OK]
:test/succeed.test:2: [OK]

         Test Cases  Total      
 Passed  2           2          
 Failed  0           0          
 Total   2           2          
shelltest --color --precise test/fail.test
:test/fail.test:1: [OK]
:test/fail.test:2: [OK]
:test/fail.test:3: [OK]
:test/fail.test:4: [OK]

         Test Cases  Total      
 Passed  4           4          
 Failed  0           0          
 Total   4           4 
```

Thanks!

**Comment by mouse07410** [comment]
Summary:

1. Need to have `cabal.project.local` with `jobs: 1` (or at most `jobs: 4`) entry.
2. Need to set env var `GHC_ENVIRONMENT=-`, at least for the tests.
3. Need to install `shelltestrunner` via Cabal (`cabal install shelltestrunner`).
4. Need to ensure, somehow, that remnants of previous (2.6.1?) builds do not linger around to interfere with the current build or tests.

First two - I recommend updating the repo, the third one - probably a line in README?

Did I miss anything?

**Comment by asr** [comment]
> Summary:
> 
>     1. Need to have `cabal.project.local` with `jobs: 1` (or at most `jobs: 4`) entry.

We need more tests before knowing what was the problem. I used `jobs: 4` because my machine has four cores. 

> 
>     2. Need to set env var `GHC_ENVIRONMENT=-`, at least for the tests.

I guess it is only needed if used Cabal v1-command for installing Agda upstream and Cabal v2-commands for other libraries and programs including released versions of Agda.

>     4. Need to ensure, somehow, that remnants of previous (2.6.1?) builds do not linger around to interfere with the current build or tests.

I guess you installed 2.6.1 with Cabal v2-commands.

**Comment by asr** [comment]
> the third one - probably a line in README?

Patches to [HACKING.md](https://github.com/agda/agda/blob/48423ad9bc224d30c57fe766395c5a4aadbba702/HACKING.md) are welcome.

**Comment by asr** [comment]
This issue is large enough. If you open a new issue for the `jobs` problem, I have a couple of ideas I want to test.

**Comment by mouse07410** [comment]
@asr I've done more experiments. It looks like `jobs: ` might matter when you *build* Agda, but is not relevant when you run tests.

Here's the patch I found necessary to successfully build *and* pass all the current tests. I suggest that if in your experience it doesn't break things on other platforms - you incorporate it into the `master`:
```diff
diff --git a/Makefile b/Makefile
index e030822c2..90c198103 100644
--- a/Makefile
+++ b/Makefile
@@ -39,6 +39,9 @@ AGDA_TESTS_OPTIONS ?=-i -j$(PARALLEL_TESTS)
 CABAL_INSTALL_HELPER = $(CABAL_CMD) $(CABAL_INSTALL_CMD) --disable-documentation
 STACK_INSTALL_HELPER = $(STACK_CMD) install Agda --no-haddock --system-ghc
 
+GHC_ENVIRONMENT=-
+AGDA_BIN="${HOME}/.cabal/bin/agda"
+
 # 2016-07-15. We use a different build directory in the quick
 # installation for avoiding recompilation (see Issue #2083 and
 # https://github.com/haskell/cabal/issues/1893).
diff --git a/src/agda-mode/Main.hs b/src/agda-mode/Main.hs
index 87ce6703d..a06fb1ac6 100644
--- a/src/agda-mode/Main.hs
+++ b/src/agda-mode/Main.hs
@@ -231,7 +231,6 @@ compileElispFiles = do
                       , "--batch"
                       , "--eval"
                       , "(progn \
-                           \(setq byte-compile-error-on-warn t) \
                            \(byte-compile-disable-warning 'cl-functions) \
                            \(batch-byte-compile))"
                       , f
```

**Comment by asr** [comment]
> ```diff
> +GHC_ENVIRONMENT=-
> +AGDA_BIN="${HOME}/.cabal/bin/agda"
> ```

You shouldn't use `agda` from the `cabal/bin` directory but from the build directory.

> ```diff
>  \(setq byte-compile-error-on-warn t)
> ```

It is not necessary remove the above line (but we need to fix https://github.com/agda/agda/issues/4610).

**Comment by mouse07410** [comment]
> You shouldn't use `agda` from the `~/.cabal/bin` directory but from the build directory.

A good point. I'll try that and report back.

>  ```lisp
> \(setq byte-compile-error-on-warn t)
> ```
> It is not necessary to remove the above line (but we need to fix #4610)

Well, it is necessary for as long as #4610 remains un-fixed.

### Update
I'm trying to figure whether allowing large number of parallel jobs is a only problem during the  build/installation of the Agda dependencies. Reason to suspect that - initially my build failed (`jobs: 20`). I limited it to 1, managed to build, got through several smaller issues, now (with all the dependencies installed) it rebuilds Agda itself and runs tests fine (with the above patches) with `jobs: 20` with no sweat.

**Comment by mouse07410** [comment]
Current Makefile:
```diff
diff --git a/Makefile b/Makefile
index e030822c2..f8428a2b4 100644
--- a/Makefile
+++ b/Makefile
@@ -39,6 +39,9 @@ AGDA_TESTS_OPTIONS ?=-i -j$(PARALLEL_TESTS)
 CABAL_INSTALL_HELPER = $(CABAL_CMD) $(CABAL_INSTALL_CMD) --disable-documentation
 STACK_INSTALL_HELPER = $(STACK_CMD) install Agda --no-haddock --system-ghc
 
+GHC_ENVIRONMENT=-
+AGDA_BIN="$(TOP)/dist-$(VERSION)/build/Agda/agda"
+
 # 2016-07-15. We use a different build directory in the quick
 # installation for avoiding recompilation (see Issue #2083 and
 # https://github.com/haskell/cabal/issues/1893).
```

Test shows that `AGDA_BIN` **must** be set to an **absolute path**, otherwise this is the result:
```
Testing vfl/Typechecker.agda... 
ok
======================================================================
======================= Standard library test ========================
======================================================================
bash: ./dist-2.6.2/build/Agda/agda: No such file or directory

real	0m0.001s
user	0m0.000s
sys	0m0.001s
make: *** [std-lib-test] Error 127

real	37m49.115s
user	50m18.227s
sys	13m39.814s
$ ll dist-2.6.2/build/Agda/agda
-rwxr-xr-x  1 ur20980  Users  22140 Jul  6 20:31 dist-2.6.2/build/Agda/agda*
```

**Comment by asr** [comment]
> ```diff
> +AGDA_BIN="$(TOP)/dist-$(VERSION)/build/Agda/agda"
> ```

This is wrong. Replace `Agda` by `agda`. Why did you need to define `AGDA_BIN`?

**Comment by mouse07410** [comment]
> This is wrong

Yes I noticed.

> Replace `Agda` by `agda`.

No, that does/did not help.

> Why did you need to define `AGDA_BIN`?

Two possibilities. One - I've nothing better to do with my time than messing with somebody else's working code in languages I do not have sufficient expertise with. Two - the code/Makefile in the repo did not work properly, and the only way to actually get (all) the tests to pass (on my platform) was setting `AGDA_BIN` env var to something reasonable. Take your pick. *`<not-quite-a-smile>`*

My experiments conclusively proved that (at least on my MacOS platform) `AGDA_BIN` must be an absolute path. For otherwise whatever drives the tests fails to correctly locate the just-built `agda` executable, and consequently fails with `Could not find module ‘Agda.Version’` (I made darn certain that there's no remnants of an old installation).

Since I always did `make install` (or `make install-bin`) *before* `make test`, the simple `AGDA_BIN=$HOME/.cabal/bin/agda` worked for me. You correctly pointed out that in general `AGDA_BIN` should point at the executable built by this process, rather than the one that got to `~/.cabal/bin` who knows when. It took me some efforts to arrive at the expression that both gives the absolute path *and* refers to `$(TOP)/dist-$(VERSION)/build/Agda/agda`.

Here's the current working patch (we already discussed `agda-mode/Main.hs` and #4610). Please feel free to do with it as you like (*the sensible thing IMHO would be to apply it*).

```diff
diff --git a/Makefile b/Makefile
index e030822c2..2dfb30835 100644
--- a/Makefile
+++ b/Makefile
@@ -39,6 +39,11 @@ AGDA_TESTS_OPTIONS ?=-i -j$(PARALLEL_TESTS)
 CABAL_INSTALL_HELPER = $(CABAL_CMD) $(CABAL_INSTALL_CMD) --disable-documentation
 STACK_INSTALL_HELPER = $(STACK_CMD) install Agda --no-haddock --system-ghc
 
+GHC_ENVIRONMENT=-
+mkfile_path := $(abspath $(lastword $(MAKEFILE_LIST)))
+mkfile_dir := $(dir $(mkfile_path))
+AGDA_BIN="$(mkfile_dir)/../dist-$(VERSION)/build/Agda/agda"
+
 # 2016-07-15. We use a different build directory in the quick
 # installation for avoiding recompilation (see Issue #2083 and
 # https://github.com/haskell/cabal/issues/1893).
diff --git a/src/agda-mode/Main.hs b/src/agda-mode/Main.hs
index 87ce6703d..a06fb1ac6 100644
--- a/src/agda-mode/Main.hs
+++ b/src/agda-mode/Main.hs
@@ -231,7 +231,6 @@ compileElispFiles = do
                       , "--batch"
                       , "--eval"
                       , "(progn \
-                           \(setq byte-compile-error-on-warn t) \
                            \(byte-compile-disable-warning 'cl-functions) \
                            \(batch-byte-compile))"
                       , f
```

**Comment by mouse07410** [comment]
OK, more feedback. It looks that no matter where in the Makefile I put the `GHC_ENVIRONMENT=-` statement, it is not sufficient. There always pops one or more of the test classes that doesn't get this setting.

It looks like the only sure way wrt. `GHC_ENVIRONMENT` is doing something like
```sh
$ GHC_ENVIRONMENT=- make test
```
or such.

Sorry.  

My other points stand.
